src/ZEO/asyncio/client.py

```diff
@@ -132,60 +144,56 @@ def pop_futures(self):
         self.futures.clear()
         return futures
 
-    def protocol_factory(self):
-        return self
-
     def connect(self):
         if isinstance(self.addr, tuple):
             host, port = self.addr
-            cr = self.loop.create_connection(
+            cr = lambda: self.loop.create_connection(  # noqa: E731
                 self.protocol_factory, host or '127.0.0.1', port,
                 ssl=self.ssl, server_hostname=self.ssl_server_hostname)
         else:
-            cr = self.loop.create_unix_connection(
+            cr = lambda: self.loop.create_unix_connection(  # noqa: E731
                 self.protocol_factory, self.addr, ssl=self.ssl)
 
-        # an ``asyncio.Future``
-        self._connecting = cr = asyncio.ensure_future(cr, loop=self.loop)
-
-        @cr.add_done_callback
-        def done_connecting(future):
-            if future.cancelled():
-                logger.info("Connection to %r cancelled", self.addr)
-            elif future.exception() is not None:
-                logger.info("Connection to %r failed, %s",
-                            self.addr, future.exception())
-            else:
-                return
-
-            # keep trying
-            if not self.closed:
+        @coroutine
```

Comment:
`@coroutine` is deprecated in Python 3.9 and removed in Python 3.11 (see https://docs.python.org/3.9/library/asyncio-task.html#asyncio-generator-based-coro)

Reply:
This is not `asyncio.coroutine` - this `@coroutine` comes from our own implementation from here:

https://github.com/zopefoundation/ZEO/blob/0d0e16ab4490adf6aa3c8549b634b89baee83f4e/src/ZEO/asyncio/futures.py#L452-L473